src/coprocessor/dag/batch/executors/limit.rs

```diff
@@ -0,0 +1,306 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use tipb::expression::FieldType;
+
+use crate::coprocessor::dag::batch::interface::*;
+use crate::coprocessor::Result;
+
+pub struct BatchLimitExecutor<Src: BatchExecutor, C: ExecSummaryCollector> {
+    src: Src,
+    remaining_rows: usize,
+    summary_collector: C,
+}
+
+impl<Src: BatchExecutor, C: ExecSummaryCollector> BatchLimitExecutor<Src, C> {
+    pub fn new(src: Src, limit: usize, summary_collector: C) -> Result<Self> {
+        if limit == 0 {
+            return Err(box_err!("limit should not be zero"));
+        }
+        Ok(Self {
+            src,
+            remaining_rows: limit,
+            summary_collector,
+        })
+    }
+}
+
+impl<Src: BatchExecutor, C: ExecSummaryCollector> BatchExecutor for BatchLimitExecutor<Src, C> {
+    #[inline]
+    fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+        let timer = self.summary_collector.start_record_duration();
+        self.summary_collector.inc_iterations();
+
+        let mut result = self.src.next_batch(expect_rows);
+        if result.data.rows_len() < self.remaining_rows {
```

Comment:
It's better to use a variable for `result.data.rows_len()`.

Reply:
then we need to change it in `else`, maybe not a good idea.

Reply:
But we can avoid calling this function twice in `if`.